flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -0,0 +1,1448 @@
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(1, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 2, and no value will be removed
+		referenceStates.get(1L).put(1, "1");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot2, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot2.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(3, stateMap.getHighestFinishedSnapshotVersion());
+
+		// add node to pruning set to prevent snapshot4 to prune
+		stateMap.getPruningValueNodes().add(node);
+		// complete snapshot 4, and no value will be removed
+		referenceStates.get(1L).put(1, "3");
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot4, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot4.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(4, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(4, stateMap.getHighestFinishedSnapshotVersion());
+
+		stateMap.getPruningValueNodes().remove(node);
+
+		// complete snapshot 5, v1 and v2 will be removed
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot5, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot5.release();
+		referenceValues.remove(referenceValues.size() - 1);
+		referenceValues.remove(referenceValues.size() - 1);
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(2, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(5, stateMap.getHighestFinishedSnapshotVersion());
+
+		// complete snapshot 6, no value will be removed
+		verifySnapshotWithoutTransform(
+			referenceStates, snapshot6, keySerializer, namespaceSerializer, stateSerializer);
+		snapshot6.release();
+		assertEquals(referenceValues, getAllValuesOfNode(stateMap, spaceAllocator, node));
+		assertEquals(2, spaceAllocator.getTotalSpaceNumber());
+		assertEquals(6, stateMap.getHighestFinishedSnapshotVersion());
+
+		assertEquals("3", stateMap.removeAndGetOld(1, 1L));
+		assertEquals(0 , stateMap.size());
+		assertEquals(0, spaceAllocator.getTotalSpaceNumber());
+
+		stateMap.close();
+	}
+
+	/**
+	 * Tests concurrent snapshots.
+	 */
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testConcurrentSnapshots() throws IOException {
+		TypeSerializer<Integer> keySerializer = IntSerializer.INSTANCE;
+		TypeSerializer<Long> namespaceSerializer = LongSerializer.INSTANCE;
+		TypeSerializer<String> stateSerializer = StringSerializer.INSTANCE;
+		// set logicalRemovedKeysRatio to 0 so that all logically removed states will be deleted when snapshot
+		CopyOnWriteSkipListStateMap<Integer, Long, String> stateMap = new CopyOnWriteSkipListStateMap<>(
+			keySerializer, namespaceSerializer, stateSerializer, spaceAllocator);
```

Comment:
This setup logic could be factored out in a separate method which could then be shared.